aten/src/ATen/quantized/Quantizer.cpp

```diff
@@ -11,11 +11,112 @@
+  const float* rd = rtensor.data<float>();
+  auto qd = reinterpret_cast<typename T::underlying*>(qtensor.data<T>());
+  fbgemm::TensorQuantizationParams qparams;
+  qparams.scale = scale;
+  qparams.zero_point = zero_point;
+  qparams.precision = std::numeric_limits<typename T::underlying>::digits;
+  fbgemm::Quantize<typename T::underlying>(/*src=*/rd,
+                             /*dst=*/qd,
+                             /*len=*/rtensor.numel(),
+                             /*qparams=*/qparams);
+  return qtensor;
+}
+
+template <typename T>
+Tensor dequantize_tensor(Tensor qtensor, Tensor rtensor, float scale, int32_t zero_point) {
+  const auto* qd = reinterpret_cast<const typename T::underlying*>(qtensor.data<T>());
+  fbgemm::TensorQuantizationParams qparams;
+  qparams.scale = scale;
+  qparams.zero_point = zero_point;
+  qparams.precision = std::numeric_limits<typename T::underlying>::digits;
+  float* rd = rtensor.data<float>();
+  fbgemm::Dequantize<typename T::underlying>(/*src=*/qd,
+                              /*dst=*/rd,
+                              /*len=*/qtensor.numel(),
+                              /*qparams=*/qparams);
+  return rtensor;
+}
+#else
+
+template <typename T>
+T quantize_val(float scale, int32_t zero_point, float value) {
+  // std::nearbyint results in nearest integer value according to the current
+  // rounding mode and the default rounding mode is rounds to even in half-way
+  // cases in most popular processor architectures like x86 and ARM. This is
+  // typically faster than an alternatives like std::round that rounds half-way
+  // cases away from zero, and can be consistent with SIMD implementations for
+  // example in x86 using _mm512_cvtps_epi32 or mm512_round_ps with
+  // _MM_FROUND_CUR_DIRECTION option that also follow the current rounding mode.
+  int32_t qvalue;
+  constexpr int32_t qmin = std::numeric_limits<typename T::underlying>::min();
+  constexpr int32_t qmax = std::numeric_limits<typename T::underlying>::max();
+  qvalue = static_cast<int32_t>(std::nearbyint(value / scale + zero_point));
+  qvalue = std::max(qvalue, qmin);
+  qvalue = std::min(qvalue, qmax);
+  return static_cast<T>(qvalue);
+}
+
+template <typename T>
+Tensor quantize_tensor(Tensor rtensor, Tensor qtensor, float scale, int32_t zero_point) {
+  const float* rdata = rtensor.data<float>();
+  auto qdata = qtensor.data<T>();
+  for (int i = 0; i < rtensor.numel(); ++i) {
+    qdata[i] = quantize_val<T>(scale, zero_point, rdata[i]);
+  }
+  return qtensor;
+}
+
+template <typename T>
+Tensor dequantize_tensor(Tensor qtensor, Tensor rtensor, float scale, int32_t zero_point) {
+  const auto* qd = qtensor.data<T>();
+  float* rd = rtensor.data<float>();
+  for (auto i = 0; i < qtensor.numel(); ++i) {
+    // We need to convert the qint8 value to float to ensure the subtraction
+    // subexpression returns a float
+    rd[i] = (static_cast<float>(qd[i].val_) - zero_point) * scale;
```

Comment:
Right now I'm just assuming everything is contiguous, we can change it later if there is a need

Reply:
Not assuming, we have .contiguous call before for the input tensor

Reply:
ok, but we already have a bunch of code for copying and and for doing type conversions.  Did you check what that code does?

Reply:
Do you mean TensorIterators?

Reply:
Since fbgemm implementation is contiguous only (and we expect that one to be enabled by default), I think it's ok to force contiguous

Reply:
It seems fine to make the output contiguous (although you still need to check in the ops themselves, because unlike say, MKLDNN, there is no guarantee you didn't do a view after getting the contiguous tensor).

But this forces the input to be contiguous too.

Reply:
And I still would like to understand how the existing copy/type conversion mechanisms fit into this.

For example, we almost certainly want memory_order support for quantized tensors, right?  If it's possible to use the same mechanism, we should get it "for free", but in this way we certainly won't.

Reply:
I think this should be a separate discussion. Also I don't have enough context on how we should handle copy/type conversion for qtensor yet.. I do plan to implement permute and view on QTensor though.